### PR TITLE
Visual basic

### PR DESCRIPTION
**Pull request by mrlacey**
**Summary of changes**
This adds support for the wizard and core engine to be able to handle templates in both C# & VB.
This includes a single page template and 2 feature templates. This is enough to test all parts of the installation experience. This PR is not intended to be enough for real world use of the templates, just enough to show it all works.
Due to the extensive nature of changes needed to get this all to work, merging other changes into this can be problematic. Hopefully, getting this code merged into the main dev branch will make things more stable and faster while the VB templates are created.

The primary purpose of this PR is to gain feedback. Got any?

Key points of interest:

- There were lots of places where the existing code assumed the template language will always be C#. These have _all_ now changed. Do not write code that assumes a specific language. There are helper methods to find this out if you need it. There is also support for "*.vb" and "*.vbproj" files where ever processing needs to know about individual file types.
- This includes the VSIX template for VB support but deliberately does not include it in the installer. (We don't want to expose this until the templates are ready for use.)
- If you want to try VB support with the VSIX, add a reference to `VBNet.UWP.VS2017.Solution` from `Installer.2017` then add the template to the vsixmanifest.
- VSEmulator does support both creating projects with C# and VB
- Right-Click>Add works for both languages.
- Many of the generation tests now take a language parameter to run for both C# & VB. Because there are only limited VB templates at the moment, these have been changed to just run for C# projects. If you want to enable this just uncomment the appropriate line from `Microsoft.Templates.Core.Language.GetAllLanguages()`
- The VB templates will use copies of XAML files from their C# equivalents. There is an issue for doing this directly with localized files as they will require some modifications, as noted in #832
- The way fixtures are created and test setup is done has had to change to allow for support of both languages. Because the last test may have been run for a different language the templates have to be reinitialized for the language of the current test. This does add some overhead but #497 can, hopefully, improve this.
- There is a "known" bug in the Suspend&Resume template. (Will be fixed in #863 if not before) : Fixed Aug7

-Which issue does this PR relate to?
#371 

-Anything that requires particular review or attention?
see above

-Do all automated tests pass?
Yes

-Have automated tests been added for new features?
Yes (lots) and many existing tests updated to run for both C# and VB versions of templates.

-If you've changed the UI are you including screenshots to show the changes
No UI changes

-Have you raised issues for any needed follow-on work?
#863 has been raised for implementing the templates in VB. The expectation/intention is that C# will remain the primary language for creating templates. Additions or modifications will then be copied/ported to VB. All work on the wizard and creating templates will continue exclusively in C#.

-Have docs been updated?
Not yet. Lots needed. Will be done with #863

-If breaking changes or different ways of doing things have been introduced, have they been communicated widely.
That's what this is for. (Aiming to show the differences, not just talk about them.)

### CODE REVIEW

**Comment by crutkas** [comment]
i hate you right now for this PR :)

**Comment by NatoBoram** [comment]
I was all excited to use Windows Template Studio but then I realized it didn't include Visual Basic. I can't wait for this to go live!

**Comment by crutkas** [comment]
this is one part of a few to enable it @NatoBoram

File: code/src/UI/Controls/CodeViewer.cs [hunk 1]
```diff
@@ -90,14 +90,18 @@ private void UpdateCodeView(Func<string, string> updateTextAction, string origin
         private string GetLanguage(string filePath)
         {
             string extension = Path.GetExtension(filePath);
-            if (extension == ".xaml" || extension == ".csproj" || extension == ".appxmanifest" || extension == ".resw" || extension == ".xml")
+            if (extension == ".xaml" || extension == ".csproj" || extension == ".vbproj" || extension == ".appxmanifest" || extension == ".resw" || extension == ".xml")
```
Comment (crutkas) [inline code/src/UI/Controls/CodeViewer.cs]: feels like a switch statement

Comment (crutkas) [inline code/src/UI/Controls/CodeViewer.cs]: IMO log new bug against this, FYI

File: code/test/Core.Test/Core.Test.csproj [hunk 2]
```diff
@@ -113,18 +102,17 @@
...
     <Reference Include="System.Xml.Linq" />
-    <Reference Include="xunit.abstractions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\xunit.abstractions.2.0.1\lib\net35\xunit.abstractions.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="xunit.abstractions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c">
+      <HintPath>..\..\packages\xunit.abstractions.2.0.0\lib\net35\xunit.abstractions.dll</HintPath>
     </Reference>
-    <Reference Include="xunit.assert, Version=2.2.0.3545, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\xunit.assert.2.2.0\lib\netstandard1.1\xunit.assert.dll</HintPath>
+    <Reference Include="xunit.assert, Version=2.1.0.3179, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c">
```
Comment (crutkas) [inline code/test/Core.Test/Core.Test.csproj]: Per above, why downgrating to 2.1?

File: code/src/Core/Language.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Templates.Core
+{
+    public static class Language
```
Comment (crutkas) [inline code/src/Core/Language.cs]: should this be "ProgrammingLanguages",  I think it should be a plural due to Colors.  https://docs.microsoft.com/en-us/uwp/api/windows.ui.colors

File: code/src/UI/Generation/UserSelection.cs [hunk 4]
```diff
@@ -22,6 +22,7 @@ public class UserSelection
         public string ProjectType { get; set; }
         public string Framework { get; set; }
         public string HomeName { get; set; }
+        public string Language { get; set; }
```
Comment (mrlacey) [inline code/src/UI/Generation/UserSelection.cs]: It makes it easier to compare with other places where it has to be a string and without an extra converter. Also, avoids enum with a '#' in the name.

File: code/src/Installer.2017/Commands/GenContextBootstrapService.cs [hunk 5]
```diff
@@ -47,7 +47,7 @@ public interface ISGenContextBootstrapService
 
     public interface IGenContextBootstrapService
     {
-        System.Threading.Tasks.Task GenContextInit();
+        System.Threading.Tasks.Task GenContextInit(string language);
```
Comment (mrlacey) [inline code/src/Installer.2017/Commands/GenContextBootstrapService.cs]: Added to the tidy up list as #924

File: code/src/UI/ViewModels/MainViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,444 @@
+﻿// ******************************************************************
```
Comment (mrlacey) [inline code/src/UI/ViewModels/MainViewModel.cs]: Addressed as part of 5e15682

File: code/test/Core.Test/Core.Test.csproj [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="..\..\packages\xunit.runner.visualstudio.2.2.0\build\net20\xunit.runner.visualstudio.props" Condition="Exists('..\..\packages\xunit.runner.visualstudio.2.2.0\build\net20\xunit.runner.visualstudio.props')" />
+  <Import Project="..\..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props" Condition="Exists('..\..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props')" />
```
Comment (mrlacey) [inline code/test/Core.Test/Core.Test.csproj]: I think this was just missed in a merge. it went up to 2.2 on July 13 while VB work was progressing separately. Will update and re run tests.

File: code/src/UI/Controls/CodeViewer.cs (hunk 1 above)
Comment (mrlacey) [inline code/src/UI/Controls/CodeViewer.cs]: raised as #925

File: code/src/Core/Language.cs (hunk 3 above)
Comment (mrlacey) [inline code/src/Core/Language.cs]: Will change

**Comment by crutkas** [comment]
you ready for the merge?

**Comment by mrlacey** [comment]
rerunning all tests (including the disabled VB ones) first to make sure the xunit changes don't have any unexpected consequences.

**Comment by mrlacey** [comment]
@crutkas looks good. time to press the big red button...

**Comment by crutkas** [comment]
BOOM

**Pull request merged**